### ISSUE TITLE
GraphCast: Fine-tuning (Impact: moderate, effort: high)

### ISSUE THREAD

**Issue by mnabian** [NVIDIA/physicsnemo#560]

**Comment by oublalkhalid** [comment]
Hello @mnabian,

Thank you for your excellent implementation using modulos! I was wondering if it's possible to inherit the pretrained checkpoint from the main JAX model. I'm exploring alternatives to leverage the pre-trained model for finetuning in this context.

Have there been any attempts to convert JAX model weights for use in the Torch modulos implementation? Thanks in advance! 

Best,
Koublal

**Comment by Flionay** [comment]
Looking forward to further progress on the GraphCast model, or demonstration of training results using Modulus Vs DeepMind jax.

**Comment by oublalkhalid** [comment]
Hello @mnabian,

Getting back to you after a few days of improvements. To be honest, I found that DALI’s iterator is not very efficient, so I implemented an iterable dataset directly, Please find below my benchmarking results comparing different dataloader strategies. It seems that the get_item method might not be optimized for handling the large size of the data. I’ll also suggest some modifications to the validation setup in my PR as loop goes to infinity for some step. My current version, trained on five years of data (2013-2017), achieves performance comparable to the JAX-based implementation from GraphCast (DeepMind Weather Team)!

<img width="411" alt="Image" src="https://github.com/user-attachments/assets/041f2ae1-7721-484a-ab7d-c23fec6a47d6" />

I look forward to your feedback on my PR!

Best,
khalid

**Comment by Betancourt20** [comment]
Hi @oublalkhalid,
Thank you for your graph and explanation, it looks awesome!!! 
Did you train it with static channels? I have been trying to train it with static channels, but it crashes in the fine-tuning part.
What we are running into is that during multi step rollout for finetuning, the size of our predicted vector (just the weather channels) is not the size of the input vector (static and weather). During the rollout, the predicted output is fed back in directly without reappending the static channels, which leads to model-mismatch errors. Then it crashes if static channel count > 0, use_time_of_year_index: or use_cos_zenith:  is true

Did you encounter the same error? or am I doing something wrong while launching the training?
Thanks in advance for your help ;)
Kind regards,
Julio

**Comment by oublalkhalid** [comment]
Hi @Betancourt20,

Thanks for your interesting question. Yeah, we only need to use the weather variables. The function [prepare_input](https://github.com/NVIDIA/physicsnemo/blob/4edf09fff19e8ab75a9385aaa18cc735ac5cae98/examples/weather/graphcast/train_utils.py#L21) (https://github.com/NVIDIA/physicsnemo/blob/4edf09fff19e8ab75a9385aaa18cc735ac5cae98/examples/weather/graphcast/train_utils.py#L21) adds static features and time features, so the input it builds is not limited to just the weather variables.

For training, you can simply use the output variable at **t-1** as the input for step **t** for the rollout case. However, for inference, the output variables are available. In this case, the predictions are then fed back into `prepare_input` to construct the updated `invar_cat` for the next step. This is how I handle the rollout. You can detach the outputs if you want to compute the gradient for only one step or a few steps using ``.detach()`` and so on. For example, with 40 steps, it’s not feasible to backpropagate through all of them, so it’s better to detach some steps. 

Let me know if this is clear or if you need further help running the code :) !

Best,
Khalid
